debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/ParallelSnapshotReader.java

```diff
@@ -0,0 +1,191 @@
+ * for all tables newly added to the config in parallel with a {@link BinlogReader} for all the tables previously
+ * in the config.
+ */
+public class ParallelSnapshotReader implements Reader {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private final BinlogReader oldTablesReader;
+    private final BinlogReader newTablesBinlogReader;
+    private final ChainedReader newTablesReader;
+
+    private final AtomicBoolean running = new AtomicBoolean();
+    private final AtomicReference<Runnable> uponCompletion = new AtomicReference<>();
+
+    /**
+     * Create a ParallelSnapshotReader.
+     *
+     * @param config the current connector configuration.
+     * @param noSnapshotContext The context for those tables not undergoing a snapshot.
+     * @param snapshotFilters {@link Filters} matching the tables that should be snapshotted.
+     */
+    public ParallelSnapshotReader(Configuration config,
+                                  MySqlTaskContext noSnapshotContext,
+                                  Filters snapshotFilters) {
+        AtomicBoolean oldTablesReaderNearEnd = new AtomicBoolean(false);
+        AtomicBoolean newTablesReaderNearEnd = new AtomicBoolean(false);
+        ParallelHaltingPredicate oldTablesReaderHaltingPredicate =
+            new ParallelHaltingPredicate(oldTablesReaderNearEnd, newTablesReaderNearEnd);
+        ParallelHaltingPredicate newTablesReaderHaltingPredicate =
+            new ParallelHaltingPredicate(newTablesReaderNearEnd, oldTablesReaderNearEnd);
+
+        this.oldTablesReader = new BinlogReader("oldBinlog", noSnapshotContext, oldTablesReaderHaltingPredicate);
+
+        MySqlTaskContext newTablesContext = new MySqlTaskContext(config, snapshotFilters);
+        SnapshotReader newTablesSnapshotReader = new SnapshotReader("newSnapshot", newTablesContext);
+
+        this.newTablesBinlogReader = new BinlogReader("newBinlog", newTablesContext, newTablesReaderHaltingPredicate);
+        this.newTablesReader = new ChainedReader().add(newTablesSnapshotReader).add(newTablesBinlogReader);
+
+    }
+
+    // for testing purposes
+    /*package private*/ ParallelSnapshotReader(BinlogReader oldTablesBinlogReader,
+                                               SnapshotReader newTablesSnapshotReader,
+                                               BinlogReader newTablesBinlogReader) {
+        this.oldTablesReader = oldTablesBinlogReader;
+        this.newTablesBinlogReader = newTablesBinlogReader;
+        this.newTablesReader = new ChainedReader().add(newTablesSnapshotReader).add(newTablesBinlogReader);
+    }
+
+    /**
+     * Create and return a {@link ReconcilingBinlogReader} for the two binlog readers contained in this
+     * ParallelSnapshotReader.
+     * @return a {@link ReconcilingBinlogReader}
+     */
+    public ReconcilingBinlogReader createReconcillingBinlogReader() {
+        return new ReconcilingBinlogReader(oldTablesReader, newTablesBinlogReader);
```

Comment:
The reconciling binlog reader needs access to the final offset, which is in the binlog reader and not the overall chained reader. I suppose one way around this would be some way of accessing the final reader in the chained reader? But I like that even less.